### PR TITLE
Add ability to update the friend assembly set of an assembly with Reflection Emit

### PR DESCRIPTION
**Pull request by davidwrighton**
- Add routine/lock for updating friend assemblies
- Unify all assembly handling custom attribute updates in the runtime
  - This adjusts the handling of the DebuggableAttribute, which in the past would have worked even if it was applied to anything via the CustomAttributeBuilder apis

Fixes the runtime level changes needed to address issues #43685 and #30917

### CODE REVIEW

File: src/coreclr/vm/comdynamic.cpp [hunk 1]
```diff
@@ -910,5 +887,64 @@ void QCALLTYPE COMDynamicWrite::DefineCustomAttribute(QCall::ModuleHandle pModul
         }
     }
 
+    // Debuggable attribute processing
```
Comment (jkotas) [inline src/coreclr/vm/comdynamic.cpp:890]: Copy&paste mistake

File: src/coreclr/vm/comdynamic.cpp [hunk 2]
```diff
@@ -910,5 +887,64 @@ void QCALLTYPE COMDynamicWrite::DefineCustomAttribute(QCall::ModuleHandle pModul
...
+                token,
+                conTok,
+                pBlob,
+                cbBlob,
+                &retToken);
+    }
+
+    if (FAILED(hr))
+    {
+        // See if the metadata engine gave us any error information.
```
Comment (jkotas) [inline src/coreclr/vm/comdynamic.cpp:932]: Does the metadata engine ever set IErrorInfo?

(I know that this code was there even before your change.)

File: src/coreclr/vm/assembly.hpp [hunk 3]
```diff
@@ -557,6 +558,12 @@ class Assembly
     void CacheManifestFiles();
 
     void CacheFriendAssemblyInfo();
+#ifndef DACCESS_COMPILE
+    ReleaseHolder<FriendAssemblyDescriptor> GetFriendAssemblyInfo(bool *pfNoFriendAssemblies);
```
Comment (jkotas) [inline src/coreclr/vm/assembly.hpp:562]: Are the copy constructors, etc. on ReleaseHolder setup correctly to make it safe to use as return value?

File: src/coreclr/vm/assembly.cpp [hunk 4]
```diff
@@ -1234,68 +1239,125 @@ void Assembly::CacheFriendAssemblyInfo()
...
+        ReleaseHolder<FriendAssemblyDescriptor> pFriendAssemblies = FriendAssemblyDescriptor::CreateFriendAssemblyDescriptor(this->GetManifestFile());
+        _ASSERTE(pFriendAssemblies != NULL);
+
+        CrstHolder friendDescriptorLock(&g_friendAssembliesCrst);
+
+        if (m_pFriendAssemblyDescriptor == NULL)
         {
-            pFriendAssemblies = NO_FRIEND_ASSEMBLIES_MARKER;
+            m_pFriendAssemblyDescriptor = pFriendAssemblies.Extract();
+            pFriendAssemblies.SuppressRelease();
```
Comment (jkotas) [inline src/coreclr/vm/assembly.cpp:1250]: `Extract` includes `SuppressRelease`. This should not be needed.

File: src/coreclr/vm/assembly.hpp (hunk 3 above)
Comment (jkotas) [reply src/coreclr/vm/assembly.hpp:562]: (We typically stay away from returning Holder types.)

File: src/coreclr/vm/comdynamic.cpp (hunk 2 above)
Comment (davidwrighton) [reply src/coreclr/vm/comdynamic.cpp:932]: I'll check. As you say, this code is old and has relics of a bygone era in it..

File: src/coreclr/vm/assembly.hpp (hunk 3 above)
Comment (davidwrighton) [reply src/coreclr/vm/assembly.hpp:562]: At some point the ReleaseHolder type I'm using was updated to have the appropriate move constructor to make this work safely. I've verified that with the new test case, the type is freed at the appropriate moment, and not earler, or later. In some work I've done where I could use tools like unique_ptr, I have found the pattern of allowing such returns to be a nice upgrade to readability.

File: src/coreclr/vm/comdynamic.cpp (hunk 2 above)
Comment (davidwrighton) [reply src/coreclr/vm/comdynamic.cpp:932]: Nope. IErrorInfo isn't set by anything in metadata. I'll get rid of this stuff.

File: src/coreclr/vm/comdynamic.cpp [hunk 5]
```diff
@@ -910,5 +887,64 @@ void QCALLTYPE COMDynamicWrite::DefineCustomAttribute(QCall::ModuleHandle pModul
...
+    // Debuggable attribute processing
+    if (((strcmp(szNamespace, FRIEND_ASSEMBLY_TYPE_NAMESPACE) == 0) && (strcmp(szName, FRIEND_ASSEMBLY_TYPE_NAME) == 0)) ||
+        ((strcmp(szNamespace, SUBJECT_ASSEMBLY_TYPE_NAMESPACE) == 0) && (strcmp(szName, SUBJECT_ASSEMBLY_TYPE_NAME) == 0)))
+    {
+        Assembly* pAssembly = pModule->GetAssembly();
+        pAssembly->UpdateCachedFriendAssemblyInfo();
+    }
+}
+
+void QCALLTYPE COMDynamicWrite::DefineCustomAttribute(QCall::ModuleHandle pModule, INT32 token, INT32 conTok, LPCBYTE pBlob, INT32 cbBlob, BOOL toDisk)
```
Comment (jkotas) [inline src/coreclr/vm/comdynamic.cpp:899]: `toDisk` argument is always false and can be deleted too.

Comment (jkotas) [reply src/coreclr/vm/comdynamic.cpp:899]: `GetOnDiskEmitter` and related code can be deleted too

Comment (davidwrighton) [reply src/coreclr/vm/comdynamic.cpp]: Good point... all that code is gone now too.

File: src/coreclr/vm/assembly.cpp [hunk 6]
```diff
@@ -1234,68 +1239,123 @@ void Assembly::CacheFriendAssemblyInfo()
...
             }
         }
+
+        // Initialize this here to avoid calling Release on the previous value of pOldFriendAssemblyDescriptor while holding the lock
+        pOldFriendAssemblyDescriptor = pFriendAssemblyDescriptorNextLoop;
     }
-} // void Assembly::CacheFriendAssemblyInfo()
+}
+
+ReleaseHolder<FriendAssemblyDescriptor> Assembly::GetFriendAssemblyInfo(bool *pfNoFriendAssemblies)
```
Comment (jkotas) [inline src/coreclr/vm/assembly.cpp:1303]: `pfNoFriendAssemblies` is not used in the method at all.

**Review by jkotas — APPROVED** [review]

**Pull request merged**